### PR TITLE
fix(python): handle pip options, hashes, and line continuations in requirements.txt

### PR DESCRIPTION
**Pull request by a-oren**
## Summary
- Adds `preprocessRequirementsLines()` to `PythonControllerBase` that properly handles all requirements.txt line types before dependency resolution
- Fixes parsing errors when requirements.txt contains `--extra-index-url`, `--hash`, line continuations (`\`), and other non-package lines
- Handles: pip options, inline options (--hash, --config-settings), line continuations, PEP 508 direct references (`name @ url`), bare URLs, VCS URLs, and local paths

Fixes [TC-4527](https://redhat.atlassian.net/browse/TC-4527)
Related: https://github.com/fabric8-analytics/fabric8-analytics-vscode-extension/issues/843

## Test plan
- [x] 13 unit tests covering all requirements.txt line types
- [x] Existing `getIgnoredDependencies_strips_environment_markers` test passes
- [ ] Manual verification with requirements.txt files containing `--extra-index-url` and line continuations

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[TC-4527]: https://redhat.atlassian.net/browse/TC-4527?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Improve Python requirements.txt preprocessing to robustly handle pip options, hashes, line continuations, and non-package lines before dependency resolution.

Bug Fixes:
- Fix handling of requirements.txt entries that include pip options, hashes, line continuations, direct references, bare URLs, VCS URLs, and local paths so they no longer break dependency parsing and installation.

Enhancements:
- Introduce a reusable preprocessing routine for requirements.txt lines and apply it across Python dependency resolution and ignored-dependency detection to ensure consistent parsing behavior.

Tests:
- Add unit tests covering preprocessing of all supported requirements.txt line variants, including pip options, hashes, line continuations, direct references, URLs, local paths, and mixed scenarios.

### CODE REVIEW

**Comment by a-oren** [comment]
@sourcery-ai review

File: src/test/java/io/github/guacsec/trustifyda/utils/PythonControllerBaseTest.java [hunk 1]
```diff
@@ -2043,4 +2043,163 @@ void when_spliting_pip_show_dep_with_license() {
...
+  @Test
+  void preprocessRequirementsLines_filters_short_pip_options() {
+    List<String> input =
+        List.of("-r other-requirements.txt", "-c constraints.txt", "numpy==1.24.0");
+    List<String> result = PythonControllerBase.preprocessRequirementsLines(input);
+    assertEquals(List.of("numpy==1.24.0"), result);
+  }
+
+  @Test
+  void preprocessRequirementsLines_strips_hashes() {
```
Comment (SourceryAI) [inline src/test/java/io/github/guacsec/trustifyda/utils/PythonControllerBaseTest.java:2068]: **suggestion (testing):** Add a test for pip option lines with leading whitespace

Since `preprocessRequirementsLines` trims each line before checking `line.startsWith("-")`, lines like `"   --extra-index-url ..."` or `"  -c constraints.txt"` are also treated as pip options and filtered out. Please add a test case with leading whitespace before these options to capture this behavior and protect against regressions in the trimming logic.

```suggestion
  @Test
  void preprocessRequirementsLines_filters_short_pip_options() {
    List<String> input =
        List.of("-r other-requirements.txt", "-c constraints.txt", "numpy==1.24.0");
    List<String> result = PythonControllerBase.preprocessRequirementsLines(input);
    assertEquals(List.of("numpy==1.24.0"), result);
  }

  @Test
  void preprocessRequirementsLines_filters_pip_options_with_leading_whitespace() {
    List<String> input =
        List.of(
            "   --extra-index-url https://pypi.example.com/simple",
            "  -c constraints.txt",
            "requests==2.28.0");
    List<String> result = PythonControllerBase.preprocessRequirementsLines(input);
    assertEquals(List.of("requests==2.28.0"), result);
  }

  @Test
  void preprocessRequirementsLines_strips_hashes() {
```

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 1 issue, and left some high level feedback:

- In preprocessRequirementsLines, the continuation-joining loop appends stripped text for continued lines but the original (unstripped) line for the final segment, which reintroduces trailing whitespace compared to the previous behavior where lines were trimmed; consider consistently using the stripped variant to avoid subtle whitespace-dependent issues.
- The getIgnoredDependencies path still processes raw split lines (now via "\\R") without using preprocessRequirementsLines, so ignore-pattern extraction may behave differently from dependency resolution for lines with pip options, hashes, or continuations; consider reusing the same preprocessing there for consistency.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In preprocessRequirementsLines, the continuation-joining loop appends stripped text for continued lines but the original (unstripped) line for the final segment, which reintroduces trailing whitespace compared to the previous behavior where lines were trimmed; consider consistently using the stripped variant to avoid subtle whitespace-dependent issues.
- The getIgnoredDependencies path still processes raw split lines (now via "\\R") without using preprocessRequirementsLines, so ignore-pattern extraction may behave differently from dependency resolution for lines with pip options, hashes, or continuations; consider reusing the same preprocessing there for consistency.

## Individual Comments

### Comment 1
<location path="src/test/java/io/github/guacsec/trustifyda/utils/PythonControllerBaseTest.java" line_range="2059-2068" />
<code_context>
             + "Required-by: cycler, gensim, gTTS, python-dateutil, tweepy\n");
   }
+
+  @Test
+  void preprocessRequirementsLines_filters_extra_index_url() {
+    List<String> input =
+        List.of(
+            "--extra-index-url https://pypi.example.com/simple",
+            "requests==2.28.0",
+            "--index-url https://pypi.org/simple",
+            "flask==2.0.3");
+    List<String> result = PythonControllerBase.preprocessRequirementsLines(input);
+    assertEquals(List.of("requests==2.28.0", "flask==2.0.3"), result);
+  }
+
+  @Test
+  void preprocessRequirementsLines_filters_short_pip_options() {
+    List<String> input =
+        List.of("-r other-requirements.txt", "-c constraints.txt", "numpy==1.24.0");
</code_context>
<issue_to_address>
**suggestion (testing):** Add a test for pip option lines with leading whitespace

Since `preprocessRequirementsLines` trims each line before checking `line.startsWith("-")`, lines like `"   --extra-index-url ..."` or `"  -c constraints.txt"` are also treated as pip options and filtered out. Please add a test case with leading whitespace before these options to capture this behavior and protect against regressions in the trimming logic.

```suggestion
  @Test
  void preprocessRequirementsLines_filters_short_pip_options() {
    List<String> input =
        List.of("-r other-requirements.txt", "-c constraints.txt", "numpy==1.24.0");
    List<String> result = PythonControllerBase.preprocessRequirementsLines(input);
    assertEquals(List.of("numpy==1.24.0"), result);
  }

  @Test
  void preprocessRequirementsLines_filters_pip_options_with_leading_whitespace() {
    List<String> input =
        List.of(
            "   --extra-index-url https://pypi.example.com/simple",
            "  -c constraints.txt",
            "requests==2.28.0");
    List<String> result = PythonControllerBase.preprocessRequirementsLines(input);
    assertEquals(List.of("requests==2.28.0"), result);
  }

  @Test
  void preprocessRequirementsLines_strips_hashes() {
```
</issue_to_address>
~~~

</details>

***

Hi @a-oren! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>

**Review by ruromero — APPROVED** [review]
LGTM
